docs/proposals/workflow.md

```diff
@@ -0,0 +1,281 @@
+[#17305](https://github.com/kubernetes/kubernetes/pull/17305): an
+_initializer_ is a dynamically registered object which permits to select a custom controller
+to be applied to a resource. The controller verifies the dependencies.
+The controller checks are applied before the resource is created (even API validated).
+Using a proper controller one may defer creation of the resource until prerequisites
+are satisfied. Even if not completed [#17305](https://github.com/kubernetes/kubernetes/pull/17305)
+already introduces a _dependency_ concept
+([see this comment](https://github.com/kubernetes/kubernetes/pull/17305#discussion_r45007826))
+which could be reused to implement `Workflow`. In
+[#1899](https://github.com/kubernetes/kubernetes/issues/1899):
+some use-cases to wait for specific conditions (`complete`, `ready`) are presented.
+
+
+### Detecting run to completion
+
+To detect run to completion for the resource inside the graph the resource needs to implement
+in `status` the slice of `condition`s. [See](../../docs/devel/api-conventions.md#objects)
+and [#7856](https://github.com/kubernetes/kubernetes/issues/7856).
+
+### Workflow
+
+A new resource will be introduced in the API. A `Workflow` is a graph.
+In the simplest case it's a graph of `Job`s but it can also
+be a graph of other entities (for example cross-cluster objects or other `Workflow`s).
+
+
+```go
+// Workflow is a directed acyclic graph
+type Workflow struct {
+    unversioned.TypeMeta `json:",inline"`
+
+    // Standard object's metadata.
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata.
+    api.ObjectMeta `json:"metadata,omitempty"`
+
+    // Spec defines the expected behavior of a Workflow. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
+    Spec WorkflowSpec `json:"spec,omitempty"`
+
+    // Status represents the current status of the Workflow. More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#spec-and-status.
+    Status WorkflowStatus `json:"status,omitempty"`
+}
+
+// WorkflowList implements list of Workflow.
+type WorkflowList struct {
+    unversioned.TypeMeta `json:",inline"`
+
+    // Standard list metadata
+    // More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
+    unversioned.ListMeta `json:"metadata,omitempty"`
+
+    // Items is the list of Workflow
+    Items []Workflow `json:"items"`
+}
+```
+
+
+#### `WorkflowSpec`
+
+```go
+// WorkflowSpec contains Workflow specification
+type WorkflowSpec struct {
+    // Optional duration in seconds relative to the startTime that the job may be active
+    // before the system tries to terminate it; value must be positive integer
+    ActiveDeadlineSeconds *int64 `json:"activeDeadlineSeconds,omitempty"`
+
+    // Steps contains the vertices of the workflow graph. The key of the map is a string
+    // to uniquely identify the step. Steps order is defined by their dependencies.
```

Comment:
About your question on JobA and JobB dependent on JobC... Sure both JobA and JobB will be started at the same time. I would say that granting only one step active is a wrong idea (and more difficult to implement)